### PR TITLE
GH-2322 repository settings for changing sail configuration

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: # 

Briefly describe the changes proposed in this PR:

---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

File: core/repository/api/src/main/java/org/eclipse/rdf4j/repository/config/RepositorySettings.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		this.settings = new HashMap<>(settings);
+	}
+
+	public RepositorySettings set(IRI iri, Literal value) {
+
+		RepositorySettings repositorySettings = new RepositorySettings(settings);
+		repositorySettings.settings.put(iri, value);
+
+		return repositorySettings;
+	}
```
Comment (hmottestad) [inline core/repository/api/src/main/java/org/eclipse/rdf4j/repository/config/RepositorySettings.java:23]: immutable

File: core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepository.java [hunk 2]
```diff
@@ -374,4 +376,16 @@ boolean useCompatibleMode() throws RepositoryException {
...
+	public RepositorySettings getSettings() {
+		try (RDF4JProtocolSession client = createHTTPClient()) {
+		}
+		return new RepositorySettings(new HashMap<>());
+	}
+
+	@Override
+	public void setSettings(RepositorySettings settings) {
+		super.setSettings(settings);
+	}
```
Comment (hmottestad) [inline core/repository/http/src/main/java/org/eclipse/rdf4j/repository/http/HTTPRepository.java:390]: This is where I would retrieve the RepositorySettings from the http client.

File: core/repository/http/src/test/java/org/eclipse/rdf4j/repository/http/HttpRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+					.iterator()
+					.next();
+
+			config.remove(null, vf.createIRI("http://rdf4j.org/config/sail/shacl#validationEnabled"), null);
+			config.add((Resource) next, vf.createIRI("http://rdf4j.org/config/sail/shacl#validationEnabled"),
+					BooleanLiteral.FALSE);
+
+			httpClient.updateRepository(new RepositoryConfig());
+
+		}
```
Comment (hmottestad) [inline core/repository/http/src/test/java/org/eclipse/rdf4j/repository/http/HttpRepositoryTest.java:51]: There is support for both retrieving and updating repository config, except you can't directly retrieve a RepositoryConfig object. I actually found it fairly hard to deal with, a lot of boiler plate code. And even though the returned statement "stream" contains the sail config, there doesn't seem to be any way of writing it back.

File: core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/MemoryStore.java [hunk 4]
```diff
@@ -366,6 +373,25 @@ public ValueFactory getValueFactory() {
...
+	public void setSettings(Map<IRI, Literal> settings) {
+		settings.forEach((key, value) -> {
+
+			if (key.equals(MemoryStoreSchema.SYNC_DELAY)) {
+				setSyncDelay(value.longValue());
+			}
+
+		});
+
+	}
```
Comment (hmottestad) [inline core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/MemoryStore.java:393]: Here is part of my solution proposal. The repository layer uses an immutable RepositorySettings object, but the sail layer just uses a Map<IRI, Literal>. The map is created by the lowest level sail, and each wrapping sail can add to the map. We reuse the vocabulary from the sail config.

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java [hunk 5]
```diff
@@ -863,4 +884,79 @@ public void setDashDataShapes(boolean dashDataShapes) {
...
+			}
+			if (key.equals(SHACL_ADVANCED_FEATURES)) {
+				setShaclAdvancedFeatures(value.booleanValue());
+			}
+			if (key.equals(DASH_DATA_SHAPES)) {
+				setDashDataShapes(value.booleanValue());
+			}
+
+		});
+	}
```
Comment (hmottestad) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java:961]: Here is the ShailSail implementation aspect of the new settings approach. Looking at this I'm partly wondering if the old fashioned model based sail config could just delegate to these settings methods.

Also an important note. For settings that are unchangeable once the sail has been initialised we would either ignore them or throw an exception.

**Comment by hmottestad** [comment]
And a final note. The discoverability on this is atrocious. A user that isn't already comfortable with the sail config approach will really struggle to use the new settings methods. We should guarantee that `getSettings()` returns all possible settings (eg. false for all disabled settings), but it's still a lot harder than just `.disableValidation()`.

**Comment by hmottestad** [comment]
I would like your feedback on this approach @jeenbroekstra before I continue with the HTTP aspect (which I'm really dreading, wish we were on spring boot tbh).

File: assembly/src/main/dist/docker/Dockerfile.amd64 [hunk 6]
```diff
@@ -1,30 +1,32 @@
 FROM amd64/tomcat:8.5-jre8-alpine
 MAINTAINER Bart Hanssens (bart.hanssens@bosa.fgov.be)
 
-ARG VERSION="2.5.1"
-
 ENV JAVA_OPTS="-Xmx2g"
 ENV CATALINA_OPTS="-Dorg.eclipse.rdf4j.appdata.basedir=/var/rdf4j"
 
-RUN adduser -S tomcat 
+RUN adduser -S tomcat
```
Comment (jeenbroekstra) [inline assembly/src/main/dist/docker/Dockerfile.amd64:7]: I'm not sure why this PR touches the Dockerfile. Accidental commit?

File: core/repository/api/src/main/java/org/eclipse/rdf4j/repository/Repository.java [hunk 7]
```diff
@@ -99,12 +100,21 @@ default void init() throws RepositoryException {
...
-	public ValueFactory getValueFactory();
+	ValueFactory getValueFactory();
+
+	default RepositorySettings getSettings() {
+		throw new UnsupportedOperationException();
+	}
+
+	default void setSettings(RepositorySettings settings) {
+		throw new UnsupportedOperationException();
+	}
```
Comment (jeenbroekstra) [inline core/repository/api/src/main/java/org/eclipse/rdf4j/repository/Repository.java:118]: I don't understand why you're adding this to the Repository interface. Didn't we want a configuration option for a Repository_Connection_ ? We already have an existing configuration model for repositories and sail stacks - the use case was inspired by needing to tweak options on the fly, per connection (or even per request).

**Comment by hmottestad** [comment]
I realized that we had talked past each other. All I want at this point is some way to configure the underlying sails from the repository (and the HttpRepository). If this is already possible then I don’t want to reinvent the wheel. 

The per-connection aspect was for extending the current methods. Eg. with ‘.begin(Validation.Disabled)’ and the revalidate method (or just validate).

**Comment by jeenbroekstra** [comment]
> I realized that we had talked past each other. All I want at this point is some way to configure the underlying sails from the repository (and the HttpRepository). If this is already possible then I don’t want to reinvent the wheel.

It is, though admittedly it is not very easy to use purely from code. If we want to improve to make it easier to manipulate I'd be all for it, but ideally we'd do that as an improvement of the existing basic mechanism. 

> The per-connection aspect was for extending the current methods. Eg. with ‘.begin(Validation.Disabled)’ and the revalidate method (or just validate).

I'd still be in favor of enabling this kind of thing through a ConnectionSetting object.

If you have time, shall we try and schedule a call later this week to talk through scenarios, and get us all on the same page?

**Comment by hmottestad** [comment]
Sent you and email to coordinate a meeting.

**Comment by jeenbroekstra** [comment]
@hmottestad I believe this PR has now been made obsolete by the work on PR #2351. Or at the very least if something is still necessary here I think we agree that we need a different approach. Can this PR be closed?

**Comment by hmottestad** [comment]
Yes. It's definitely obsolete for the problem it's trying to solve. It would still be great to allow some of the sail config a to be changed at runtime, but it's not something I intend to pursue.

**Pull request closed without merging**